### PR TITLE
Unify double and decimal into number

### LINKED ISSUE #1129: Unify Integer and Decimal classes in a single Number wollok class
Natively we should use only a java.lang.BigDecimal, and show decimal points or commas depending on i18n

### PR DESCRIPTION
**Pull request by fdodino**
Fixes #1129 

Behavior is similar to previous version.

### CODE REVIEW

File: org.uqbar.project.wollok.lib/src/wollok/lang.wlk [hunk 1]
```diff
@@ -724,7 +724,8 @@ class List inherits Collection {
 		if (self.isEmpty()) 
 			throw new Exception("Illegal operation 'anyOne' on empty collection")
 		else 
-			return self.get(0.randomUpTo(self.size()))
+			//return self.get(0.randomUpTo(self.size()))
```
Comment (npasserini) [inline org.uqbar.project.wollok.lib/src/wollok/lang.wlk]: This changes the behavior of anyOne, without any discussion.
I think we should keep original behavior.

File: org.uqbar.project.wollok.lib/src/wollok/lang.wlk [hunk 2]
```diff
@@ -1055,7 +1128,7 @@ class Number {
 	 * 14.6165.roundUp(3) ==> 14.617
 	 * 5.roundUp(3) ==> 5
 	 */
-	 method roundUp(_decimals)
+	method roundUp(_decimals) native
```
Comment (npasserini) [inline org.uqbar.project.wollok.lib/src/wollok/lang.wlk]: Couldn't we just forget the "_" prefix?

File: org.uqbar.project.wollok.lib/src/wollok/lang.wlk [hunk 3]
```diff
@@ -1065,89 +1138,32 @@ class Number {
...
-	override method stringValue() native	
-	
+	 
 	/** 
 	 * Builds a Range between self and end
 	 * 
 	 * Example:
-	 * 		1..4   Answers ==> a new Range object from 1 to 4
+	 * 		1..4      Answers ==> a new Range object from 1 to 4
+	 *      1.0..4.0  Answers ==> a new Range object from 1 to 4
```
Comment (npasserini) [inline org.uqbar.project.wollok.lib/src/wollok/lang.wlk]: Given that there is now no difference between 1 and 1.0, I think that this second example only adds confusion.

File: org.uqbar.project.wollok.lib/src/wollok/lang/Range.xtend [hunk 4]
```diff
@@ -39,8 +40,12 @@ class Range extends AbstractJavaWrapper<IntegerRange> {
 	
 	def Integer validate(Object value) {
 		try {
-			return value.wollokToJava(Integer) as Integer
-		} catch (Throwable e) {
+			val castedValue = value.wollokToJava(BigDecimal) as BigDecimal
```
Comment (npasserini) [inline org.uqbar.project.wollok.lib/src/wollok/lang/Range.xtend]: I think this cast could be avoided if we add a generic return type to wollokTojava:
`def <T> T wollokToJava(Class<? extends T> type)`, isn't it?

File: org.uqbar.project.wollok.lib/src/wollok/lang/WNumber.xtend [hunk 5]
```diff
@@ -1,63 +1,165 @@
...
+
+	@NativeMessage("<")
+	def lesser(WollokObject other) { wrapped.doubleValue < other.nativeNumber.wrapped.doubleValue }
+
+	@NativeMessage("<=")
+	def lesserOrEquals(WollokObject other) { wrapped.doubleValue <= other.nativeNumber.wrapped.doubleValue }
+
+	def gcd(WollokObject other) {
+		if (!wrapped.isInteger) {
+			throw new IllegalArgumentException("gcd expects receiver to be an integer")
```
Comment (npasserini) [inline org.uqbar.project.wollok.lib/src/wollok/lang/WNumber.xtend]: I would prefer "gcd can only be applied to integer numbers".

Also this should be a pure Wollok exception, but that problem is already reported in this file, so I will create an issue to solve it later.

File: org.uqbar.project.wollok.launch/.classpath [hunk 6]
```diff
@@ -2,11 +2,11 @@
 <classpath>
```
Comment (fdodino) [inline org.uqbar.project.wollok.launch/.classpath]: !! I don't know why happened this?

File: org.uqbar.project.wollok.lib/src/wollok/lang.wlk (hunk 1 above)
Comment (fdodino) [inline org.uqbar.project.wollok.lib/src/wollok/lang.wlk]: Indeed! I was testing some error and forgot to rollback.

File: org.uqbar.project.wollok.lib/src/wollok/lang/Range.xtend (hunk 4 above)
Comment (fdodino) [inline org.uqbar.project.wollok.lib/src/wollok/lang/Range.xtend]: Will leave this for future change. Now we have

```xtend
def static Object wollokToJava(Object o, Class<?> t) {
````

**Comment by fdodino** [comment]
Well, at least this program compiles ok with constrained type system activated (issue #1146)

image

Well, @npasserini I agree there is no need to introduce right now this functionality, so we can leave this PR open.

**Comment by fdodino** [comment]
We decided with @npasserini  to deprecate this PR favoring next PR with additional features.

**Comment by fdodino** [comment]
Closed PR since I'm now working on a new one.

**Pull request closed without merging**